mwifiex: Fixes for wifi p2p and AP mode

jonas2515 opened the pull request:
Make wifi p2p actually work and fix a firmware crash when switching from station mode to AP mode.

qzed commented on drivers/net/wireless/marvell/mwifiex/cfg80211.c:
```diff
@@ -962,6 +962,138 @@ mwifiex_init_new_priv_params(struct mwifiex_private *priv,
+		case NL80211_IFTYPE_UNSPECIFIED:
+		case NL80211_IFTYPE_ADHOC:
+		case NL80211_IFTYPE_STATION:
+			adapter->curr_iface_comb.sta_intf--;
+			break;
+		case NL80211_IFTYPE_AP:
+			adapter->curr_iface_comb.uap_intf--;
+			break;
+		case NL80211_IFTYPE_P2P_CLIENT:
+		case NL80211_IFTYPE_P2P_GO:
+			adapter->curr_iface_comb.p2p_intf--;
+			break;
+		default:
+			/* This should be dead code; checked above */
+			mwifiex_dbg(adapter, ERROR, "type not supported\n");
+			break;
+		}
+	}
+
+	if (new_iftype != -1) {
+		switch (new_iftype) {
+		case NL80211_IFTYPE_UNSPECIFIED:
+		case NL80211_IFTYPE_ADHOC:
+		case NL80211_IFTYPE_STATION:
+			adapter->curr_iface_comb.sta_intf++;
+			break;
+		case NL80211_IFTYPE_AP:
+			adapter->curr_iface_comb.uap_intf++;
+			break;
+		case NL80211_IFTYPE_P2P_CLIENT:
+		case NL80211_IFTYPE_P2P_GO:
+			adapter->curr_iface_comb.p2p_intf++;
+			break;
+		default:
+			/* This should be dead code; checked above */
+			mwifiex_dbg(adapter, ERROR, "type not supported\n");
+			break;
+		}
+	}
+}
```
I think it would be cleaner to split this into two functions, one for incrementing and one for decrementing (removing the need for the special `-1` value). Also you can split out a third function that updates the counters based on an input value, i.e. `update(type, value)` and then use `update(type, +1)` and `update(type, -1)` inside some small wrappers for inc and dec (basically just for naming).

jonas2515 replied:
Went with only one helper function that gets passed `+1`/`-1` directly now, adding increase/decrease wrappers seemed a little too much IMHO.

qzed commented on drivers/net/wireless/marvell/mwifiex/cfg80211.c:
```diff
@@ -962,6 +962,138 @@ mwifiex_init_new_priv_params(struct mwifiex_private *priv,
+			return true;
+		case NL80211_IFTYPE_P2P_CLIENT:
+		case NL80211_IFTYPE_P2P_GO:
+			return adapter->curr_iface_comb.p2p_intf ==
+			       adapter->iface_limit.p2p_intf;
```
Are these checks correct? In the original code you have
```c
if (adapter->curr_iface_comb.p2p_intf ==
    adapter->iface_limit.p2p_intf) {
	mwifiex_dbg(adapter, ERROR,
		    "cannot create multiple P2P ifaces\n");
	return -1;
}
```
so shouldn't this be
```c
return adapter->curr_iface_comb.p2p_intf !=
       adapter->iface_limit.p2p_intf;
```
instead?

jonas2515 replied:
Ouch, nice catch, indeed! That's what reviews are for, aren't they ;)

jonas2515 replied:
Fixed that now

qzed reviewed (changes requested):
Can't comment on the functionality, but code looks ok. I'm not quite sure what to think of splitting all the transition checks out of the individual transition functions (I generally like to keep those things as local as possible), but I guess in this case it does simplify the transitioning code quite a bit, so I'm okay with that.

jonas2515 commented:
Alright, I've updated this to target 5.11 and added a few more fixes. IMHO the quality of this code is quite embarrassing for Marvell, it looks almost like the whole wifi p2p functionality was never tested.

qzed commented on drivers/net/wireless/marvell/mwifiex/cfg80211.c:
```diff
@@ -1214,88 +1288,65 @@ mwifiex_cfg80211_change_virtual_intf(struct wiphy *wiphy,
+		case NL80211_IFTYPE_AP:
+			return mwifiex_change_vif_to_ap(dev, curr_iftype, type,
+							params);
+		default: /* Should never get here, handled by is_vif_type_change_allowed() */
```
You could add a `WARN(true, "unchecked interface transition: %d to %d", curr_iftype, type);` here. That way if this ever happens it gets noticed ant we know what change caused the warning.

jonas2515 replied:
Good idea, went with `mwifiex_dbg` for consistency and used a goto to avoid duplicating it everywhere.

qzed replied:
I'm not entirely sure if `mwifiex_dbg` is the best function for that since it seems to use `dev_info`/`pr_info` under the hood, maybe something stronger that sticks out in dmesg would be better. On the other hand though, they essentially seem to have replicated `dev_{info,warn,err,...}` with that so I don't know what to think of that macro...

Probably better to stick with it for consistency.

qzed reviewed (approved):
One small debugability thing again, but otherwise good as usual!
